src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/template/SimpleTemplateEngineStepExecution.java

```diff
@@ -0,0 +1,107 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.pipeline.utility.steps.template;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.FilePath;
+import groovy.text.SimpleTemplateEngine;
+import groovy.text.Template;
+import org.apache.commons.io.IOUtils;
+import org.jenkinsci.plugins.pipeline.utility.steps.AbstractFileOrTextStepExecution;
+import org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.GroovySandbox;
+import org.jenkinsci.plugins.scriptsecurity.sandbox.Whitelist;
+import org.jenkinsci.plugins.scriptsecurity.sandbox.whitelists.ProxyWhitelist;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+
+import java.io.FileNotFoundException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.io.IOException;
+
+import static org.apache.commons.lang.StringUtils.isBlank;
+import static org.apache.commons.lang.StringUtils.isNotBlank;
+
+/**
+ * Execution of {@link SimpleTemplateEngineStep}.
+ *
+ * @author Martin d'Anjou
+ */
+public class SimpleTemplateEngineStepExecution extends AbstractFileOrTextStepExecution<Object> {
+    private static final long serialVersionUID = 1L;
+
+    private transient SimpleTemplateEngineStep step;
+
+    protected SimpleTemplateEngineStepExecution(@NonNull SimpleTemplateEngineStep step, @NonNull StepContext context) {
+        super(step, context);
+        this.step = step;
+    }
+
+    @Override
+    protected String doRun() throws Exception {
+        String fName = step.getDescriptor().getFunctionName();
+        if (isNotBlank(step.getFile()) && isNotBlank(step.getText())) {
+            throw new IllegalArgumentException(Messages.SimpleTemplateEngineStepExecution_tooManyArguments(fName));
+        }
+        
+        if (step.getBindings() == null) {
+            throw new IllegalArgumentException(Messages.SimpleTemplateEngineStepExecution_missingBindings(fName));
+        }
+
+        SimpleTemplateEngine engine = new SimpleTemplateEngine();
+        Template template = null;
+        if (isNotBlank(step.getFile())) {
+            FilePath f = ws.child(step.getFile());
+            if (f.exists() && !f.isDirectory()) {
+                try (InputStream is = f.read()) {
+                    template = engine.createTemplate(IOUtils.toString(is, StandardCharsets.UTF_8));
```

Comment:
I don't know but it could be that static code in the template (if that is possible) could be run here. Needs to be tested.

Reply:
Needs to be checked if it is allowed to be run if it is not running in the sandbox.
```suggestion
                    String tmpl = IOUtils.toString(is, StandardCharsets.UTF_8);
                    if (!isRunInSandBox()) {
                        ScriptApproval.get().configuring(tmpl, GroovyLanguage.get(), ApprovalContext.create().withItem(getContext().get(Item.class)), false);
                    }
                    template = engine.createTemplate(tmpl);
```

Reply:
Looking at the code for `SimpleTemplateEngine` it is in `createTemplate` that the script is parsed, so potentially some static code could run outside of the sandbox, but I am not sure.

Reply:
Yeah, it is not safe to call `createTemplate` outside of the sandbox.